src/mpid/ch4/netmod/ofi/ofi_events.c

```diff
@@ -695,7 +695,7 @@ static int am_read_event(struct fi_cq_tagged_entry *wc, MPIR_Request * dont_use_
 
 
     if (MPIDI_OFI_AMREQUEST(rreq, rdma_read_hdr)->lmt_type == MPIDI_OFI_AM_LMT_IOV) {
-        if (read_req->lmt_cntr > 0) {
+        if (!read_req->is_last) {
```

Comment:
this assumes that the read completions are ordered?

Reply:
Yes. The old code also assumes the ordered completion. The `read_req->lmt_cntr` was just a static number set at issuing time.

Reply:
Hm, so this is just broken.

Reply:
Do you mean the read completion is not ordered?

Reply:
In the non-contig path, separate reads is chained by issuing the next read from the previous completion event. We could do the same for the contig case.

Reply:
I don't think coding this as a boolean in the last read request is a good idea. We need to ensure that the parent request only completes after all reads are done. So either the counter needs to get attached to the parent request, or perhaps incorporated into the completion counter

Reply:
Alternatively, we can move `lmt_cntr` to the original request object and atomically decrement it.

Reply:
> 
> 
> I don't think coding this as a boolean in the last read request is a good idea. We need to ensure that the parent request only completes after all reads are done. So either the counter needs to get attached to the parent request, or perhaps incorporated into the completion counter

Okay, let me rewrite this into atomic counter in the parent request